### PR TITLE
16527 - check token cache endpoint before generating new service account token

### PR DESCRIPTION
**Pull request by bolyachevets**
*Issue #: https://github.com/bcgov/entity/issues/16527

*Description of changes: added a service account token caching service in gcp as a POC: https://github.com/bolyachevets/token_cache

Looking for some feedback.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-auth license (Apache 2.0).

### CODE REVIEW

**Comment by seeker25** [comment]
Interesting - why not just do a memory cache and save yourself the RTT?

**Review by seeker25 — APPROVED** [review]

**Review by seeker25 — COMMENTED** [review]
Please fix the unit tests / lints before merging

**Comment by seeker25** [comment]
Also, maybe a better idea to repoint keycloak token fetching to just use your caching service instead? Thoughts?

**Comment by bolyachevets** [comment]
> Maybe a suggestion? Repoint keycloak token fetching to just use your caching service instead (like wrapping keycloak)? Maybe through the apigee gateway? @thorwolpert -^ maybe you have some ideas?

right now it is just an 'in memory' cache, since i only had in mind caching a few service account tokens

**Comment by seeker25** [comment]
Cool whichever way you want to do it, you have my support

**Review by pwei1018 — APPROVED** [review]

**Comment by thorwolpert** [comment]
Is this for service accounts, or token caching in GCP?
if it's for caching in GCP land, why wouldn't we use memorystore (redis or memchache)?
does it have any similarities to the POC with IDIM?
https://github.com/bcgov/entity/blob/main/rfcs/rfc-3rd-party-otp.md

**Comment by bolyachevets** [comment]
> Is this for service accounts, or token caching in GCP? if it's for caching in GCP land, why wouldn't we use memorystore (redis or memchache)? does it have any similarities to the POC with IDIM? https://github.com/bcgov/entity/blob/main/rfcs/rfc-3rd-party-otp.md

the immediate usage is to cache keycloak client service account tokens used on the openshift side, e.g. in queues

**Comment by thorwolpert** [comment]
Those a re long running services, could you just use the local cache service?
like
[api/namex/services/cache.py](https://github.com/bcgov/namex/blob/main/api/namex/services/cache.py)
which uses the pallets cache project?
https://github.com/pallets-eco/cachelib

**Comment by bolyachevets** [comment]
> Those a re long running services, could you just use the local cache service? like [api/namex/services/cache.py](https://github.com/bcgov/namex/blob/main/api/namex/services/cache.py) https://github.com/bcgov/namex/blob/main/api/namex/resources/statistics/wait_time_statistics.py#L34
> 
> which uses the pallets cache project? https://github.com/pallets-eco/cachelib
> 
> Not sure if it meets your needs, but has a nice uage: `@cache.cached(timeout=14400) # cached for 4 hours`
> 
> I use cachelib to cache the JWKS, so we're not pulling the key and cert from keycloak all the time when validating the incoming JWT. Seems like it would work for caching service-acct JWTs in a similar fashion

- One of the requirements was to cache tokens outside of app context
- These are today's counts for number of times we called 'protocol/openid-connect/token' endpoint for each client.

-------------06/22/2023-------------
api-key-account-6923 : 60
api-key-account-4637-sandbox : 14
business-search-service-account : 786
cso-service-account : 1367
entity-notebook-service-account : 188
name-request-service-account : 24447
lear_sa : 40930
api-key-account-3963 : 3
name-request : 55
terraform : 8
api-key-account-4637 : 50
api-key-account-14355 : 36
ppr-service-account : 2837
api-key-account-7140 : 53
api-key-account-5401 : 4
api-key-account-13301-sandbox : 36
api-key-account-6873 : 1
api-key-account-5797-sandbox : 3
entity-service-account : 42951
ppr-document-delivery-service-account : 1043
sbc-auth-admin : 1149
api-key-account-5841 : 39
api-key-account-6833 : 10
api-key-account-5797 : 102
api-key-account-5838 : 54
namex-solr-service-account : 1963
api-key-account-6502 : 31

**Comment by thorwolpert** [comment]
which cachelib does, as it's used here to not keep generating the stat.
It's what the library and cache were designed to do.
image

**Comment by thorwolpert** [comment]
It caches across calls (contexts), https://flask-caching.readthedocs.io/en/latest/
as stated before, that's how we cache the JWKS. If you look it's not getting JWKS for every JWT across every user/session/context.

**Comment by thorwolpert** [comment]
i meant the service itself should add in cachelib or flask-caching.
you are building a backing store, and for the long running processes they don't need it.
for shortlived things, like knative containers, the backing store could be redis/mamcache, but still use cachelib/flask-cache in the service and doesn't need to know more, as it becomes a deployment config/decision

**Comment by bolyachevets** [comment]
> i meant the service itself should add in cachelib or flask-caching. you are building a backing store, and for the long running processes they don't need it. for shortlived things, like knative containers, the backing store could be redis/memcache, but still use cachelib/flask-cache in the service and doesn't need to know more, as it becomes a deployment config/decision

Just to clarify, I thought you meant that hooking up a caching library directly to RestService token fetching method would cache token values across calls (as in the most recent submission), or do you mean that I should use a caching library for the token fetching route running in gcp

**Comment by thorwolpert** [comment]
If I were adding caching to PPR or Lear, I would add cachelib or flask-caching to those services.
If they are running in OCP, I'm probably done.
If they are running in GCP, I'd change the config to point the caching store to redis or memcache.

**Comment by bolyachevets** [comment]
> memcache

I tried setting up memorystore in gcp with redis and memcached options. redis estimate is > 100$ a month, memcached is > 40$ a month.

I guess we would be using for more than just caching a couple of keycloak tokens. I did a check in cache.py to see if redis/memcached memorystores are configured.

**Comment by thorwolpert** [comment]
I was not suggesting you do that.
It should be done in the client, not as a newly created service.

**Comment by bolyachevets** [comment]
> I was not suggesting you do that. It should be done in the client, not as a newly created service.
what do you mean by doing it in the client?
e.g., to instantiate flask-caching client via memorycached backend you need to provide a list of server addresses. where would these servers come from, if not as gcp memorystore services?

**Comment by thorwolpert** [comment]
Like it's done in NameX api, auth, JWKS, etc. the cachelib is in the API.  
When running ling services in OCP, long jobs in GCP, then the simple-cache is fine. For short lived services like CloudRun, then you'd change the config in that container service to use redis/memcache. 
Building a competitor to redis/memcache or creating a new module/library for the apps isn't something we want to do.

**Comment by bolyachevets** [comment]
> cachelib

I may be not understanding this completely, but I don't think I created a competitor to redis/memcache. Looking at comments in cachelib's memcached.py you can see that it needs a memcached library to instantiate:     ```
The first argument can either be an object that resembles the API of a
    :class:`memcache.Client` or a tuple/list of server addresses. In the
    event that a tuple/list is passed, Werkzeug tries to import the best
    available memcache library.

    This cache looks into the following packages/modules to find bindings for
    memcached:

        - ``pylibmc``
        - ``google.appengine.api.memcached``
        - ``memcached``
        - ``libmc``
```
Similar for redis.

What I did is modify the cache loader file https://github.com/bcgov/sbc-auth/blob/1be82e530036052d47d6b1f2d2e93e94c951d4a7/auth-api/src/auth_api/utils/cache.py to check whether memcache/redis relevant parameters have been set up in the config. If they are, then appropriate backend would be loaded, otherwise if would stick with simple cache (which has been used in the repos you mentioned, and what is fine for the purpose of caching service account tokens as well). When I spoke with @pwei1018 , I was under the impression that we want to make sure this will work if we decide to use memcache/redis in gcp. I tried using memcache/redis in gcp, and the backend would fail to load unless proper library is available (i.e.  for memcache these would be    - ``pylibmc``
        - ``google.appengine.api.memcached``
        - ``memcached``
        - ``libmc`` ) In summary, nothing changed in the behavior of cache - it is still using simple cache, but is also robust to use mecache or redis.

**Comment by seeker25** [comment]
After closer inspection, I don't think we're reinventing the wheel here, looks like the same pathway of Namex. 

Platform folks (SSO team) are seeing a spike of keycloak activity, have notified Andriy about it - I think we should tweak this PR a bit and get it rolling. 

We might also want to create service accounts specific for each application so we can trace these down in the future. EG. LEGAL-API uses the same service account as AUTH.

File: auth-api/src/auth_api/utils/cache.py [hunk 1]
```diff
@@ -12,8 +12,23 @@
...
 # limitations under the License.
 """Bring in the common cache."""
+import os
 from flask_caching import Cache
 
+cache_servers = os.environ.get('CACHE_MEMCACHED_SERVERS')
 
-# lower case name as used by convention in most Flask apps
-cache = Cache(config={'CACHE_TYPE': 'simple'})  # pylint: disable=invalid-name
+if not cache_servers is None:
```
Comment (seeker25) [inline auth-api/src/auth_api/utils/cache.py:20]: Double negative

File: auth-api/src/auth_api/services/rest_service.py [hunk 2]
```diff
@@ -163,11 +164,13 @@ def get(endpoint, token=None,  # pylint: disable=too-many-arguments
         return response
 
     @staticmethod
+    @cache.cached()
```
Comment (seeker25) [inline auth-api/src/auth_api/services/rest_service.py:167]: needs to be tied to `CACHE_DEFAULT_TIMEOUT`?

Comment (seeker25) [reply auth-api/src/auth_api/services/rest_service.py:167]: nvm:

image

**Review by Jxio — APPROVED** [review]

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://bcregistry-account-dev--pr-2369-httbbpif.web.app

**Comment by seeker25** [comment]
Good stuff, not sure why you ran /gcbrun, only api changes

**Comment by bolyachevets** [comment]
> Good stuff, not sure why you ran /gcbrun, only api changes

for some reason gcp check was listed as one of the conditions for the green checkmark

**Comment by seeker25** [comment]
yeah we usually just ignore it haha

**Pull request merged**